akka-stream/src/main/scala/akka/stream/io/Framing.scala

```diff
@@ -0,0 +1,173 @@
+    private val firstSeparatorByte = separatorBytes.head
+    private var buffer = ByteString.empty
+    private var nextPossibleMatch = 0
+    private var finishing = false
+
+    override def onPush(chunk: ByteString, ctx: Context[ByteString]): SyncDirective = {
+      buffer ++= chunk
+      doParse(ctx)
+    }
+
+    override def onPull(ctx: Context[ByteString]): SyncDirective = {
+      doParse(ctx)
+    }
+
+    override def onUpstreamFinish(ctx: Context[ByteString]): TerminationDirective = {
+      if (buffer.nonEmpty) ctx.absorbTermination()
+      else ctx.finish()
+    }
+
+    private def tryPull(ctx: Context[ByteString]): SyncDirective = {
+      if (ctx.isFinishing) ctx.fail(new IllegalStateException(
+        "Stream finished but there was a truncated final frame in the buffer"))
+      else ctx.pull()
+    }
+
+    @tailrec
+    private def doParse(ctx: Context[ByteString]): SyncDirective = {
+      val possibleMatchPos = buffer.indexOf(firstSeparatorByte, from = nextPossibleMatch)
+      if (possibleMatchPos > maximumLineBytes)
+        ctx.fail(new IllegalStateException(s"Read ${buffer.size} bytes " +
+          s"which is more than $maximumLineBytes without seeing a line terminator"))
+      else {
+        if (possibleMatchPos == -1) {
+          // No matching character, we need to accumulate more bytes into the buffer
+          nextPossibleMatch = buffer.size
+          tryPull(ctx)
+        } else if (possibleMatchPos + separatorBytes.size > buffer.size) {
+          // We have found a possible match (we found the first character of the terminator
+          // sequence) but we don't have yet enough bytes. We remember the position to
+          // retry from next time.
+          nextPossibleMatch = possibleMatchPos
+          tryPull(ctx)
+        } else {
+          if (buffer.slice(possibleMatchPos, possibleMatchPos + separatorBytes.size)
+            == separatorBytes) {
+            // Found a match
+            val parsedFrame = buffer.slice(0, possibleMatchPos).compact
+            buffer = buffer.drop(possibleMatchPos + separatorBytes.size)
+            nextPossibleMatch -= possibleMatchPos + separatorBytes.size
+            if (ctx.isFinishing && buffer.isEmpty) ctx.pushAndFinish(parsedFrame)
+            else ctx.push(parsedFrame)
+          } else {
+            nextPossibleMatch += 1
+            doParse(ctx)
+          }
+        }
+      }
+    }
+
+  }
+
+  class LengthFieldFramingStage(
+    val lengthFieldLength: Int,
+    val lengthFieldOffset: Int,
+    val byteOrder: ByteOrder) extends PushPullStage[ByteString, ByteString] {
+    private var buffer = ByteString.empty
+    private val minimumChunkSize = lengthFieldOffset + lengthFieldLength
+    private val intDecoder: IntDecoder = byteOrder match {
+      case ByteOrder.BIG_ENDIAN    ⇒ new BigEndianCodec(lengthFieldLength)
+      case ByteOrder.LITTLE_ENDIAN ⇒ new LittleEndianCodec(lengthFieldLength)
+    }
+    private var frameSize = Int.MaxValue
```

Comment:
> will means we may do something like dropping or fail when the the maxLength of the frame is read.

But that is impossible. We dont "read", but we receive arbitrary many bytes from the upstream. That chunk might contain many valid frames not just one. If someone wants to limit the chunkSize, he can always put such stage before this one.